File: actors/tiktok-challenge-posts-scraper/src/input.ts
export const MAX_CHALLENGE_IDS = 20;
export const MAX_RESULTS_PER_CHALLENGE = 500;
export const MAX_TOTAL_RESULTS = 2_000;
export const MAX_PAGE_SIZE = 50;

export interface ChallengePostsInput {
    challenge_ids?: unknown;
    challenge_id?: unknown;
    region?: unknown;
    cursor?: unknown;
    results_per_challenge?: unknown;
    page_size?: unknown;
}

export interface ChallengePostsRequest {
    challengeId: string;
    region?: string;
    initialCursor?: string;
    resultLimit: number;
    pageSize: number;
}

function normalizeId(value: unknown): string | null {
    if (typeof value === 'number' && Number.isSafeInteger(value)) {
        value = String(value);
    }
    if (typeof value !== 'string' || !/^\d{1,100}$/.test(value.trim())) {
        return null;
    }
    return value.trim();
}

function positiveInteger(value: unknown, fallback: number, maximum: number, name: string): number {
    if (value === undefined) {
        return fallback;
    }
    if (typeof value !== 'number' || !Number.isInteger(value) || value < 1 || value > maximum) {
        throw new Error(`${name} must be an integer between 1 and ${maximum}`);
    }
    return value;
}

export function parseInput(input: ChallengePostsInput): ChallengePostsRequest[] {
    const batchValues = Array.isArray(input.challenge_ids) ? input.challenge_ids : [];
    const values = batchValues.length > 0 ? batchValues : [input.challenge_id];
    const ids = [...new Set(values.map(normalizeId).filter((id): id is string => id !== null))];

    if (ids.length === 0) {
        throw new Error('At least one numeric TikTok challenge ID is required');
    }
    if (ids.length > MAX_CHALLENGE_IDS) {
        throw new Error(`A maximum of ${MAX_CHALLENGE_IDS} challenge IDs is allowed per run`);
    }

    const resultLimit = positiveInteger(input.results_per_challenge, 100, MAX_RESULTS_PER_CHALLENGE, 'results_per_challenge');
    if (ids.length * resultLimit > MAX_TOTAL_RESULTS) {
        throw new Error(`Total requested results cannot exceed ${MAX_TOTAL_RESULTS}`);
    }
    const pageSize = Math.min(positiveInteger(input.page_size, 10, MAX_PAGE_SIZE, 'page_size'), resultLimit);

    const region = typeof input.region === 'string' && input.region.trim() !== '' ? input.region.trim().toUpperCase() : undefined;
    if (region && !/^[A-Z]{2}$/.test(region)) {
        throw new Error('region must be a two-letter country code');
    }

    let initialCursor: string | undefined;
    if (typeof input.cursor === 'string' && input.cursor.trim() !== '') {
        initialCursor = input.cursor.trim();
    } else if (typeof input.cursor === 'number' && Number.isSafeInteger(input.cursor)) {
        initialCursor = String(input.cursor);
    } else if (input.cursor !== undefined && input.cursor !== null) {
        throw new Error('cursor must be a string or safe integer');
    }

    return ids.map((challengeId) => ({ challengeId, region, initialCursor, resultLimit, pageSize }));
}

File: actors/tiktok-challenge-posts-scraper/src/input.rs
use anyhow::{bail, Result};
use serde_json::Value;
use std::collections::HashSet;

pub const MAX_CHALLENGE_IDS: usize = 20;
pub const MAX_RESULTS_PER_CHALLENGE: usize = 500;
pub const MAX_TOTAL_RESULTS: usize = 2_000;
pub const MAX_PAGE_SIZE: usize = 50;

const MAX_SAFE_INTEGER: f64 = 9_007_199_254_740_991.0;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChallengeRequest {
    pub challenge_id: String,
    pub region: Option<String>,
    pub initial_cursor: Option<String>,
    pub result_limit: usize,
    pub page_size: usize,
}

pub fn parse_input(input: &Value) -> Result<Vec<ChallengeRequest>> {
    let object = input.as_object();
    let batch_values = object
        .and_then(|object| object.get("challenge_ids"))
        .and_then(Value::as_array)
        .filter(|values| !values.is_empty());
    let values = batch_values
        .map(|values| values.iter().collect::<Vec<_>>())
        .unwrap_or_else(|| {
            object
                .and_then(|object| object.get("challenge_id"))
                .into_iter()
                .collect()
        });

    let mut unique_ids = Vec::new();
    let mut seen_ids = HashSet::new();
    for value in values {
        let Some(id) = normalize_challenge_id(value) else {
            continue;
        };
        if seen_ids.insert(id.clone()) {
            unique_ids.push(id);
        }
    }

    if unique_ids.is_empty() {
        bail!("At least one numeric TikTok challenge ID is required");
    }
    if unique_ids.len() > MAX_CHALLENGE_IDS {
        bail!("A maximum of {MAX_CHALLENGE_IDS} challenge IDs is allowed per run");
    }

    let result_limit = positive_integer(
        object.and_then(|object| object.get("results_per_challenge")),
        100,
        MAX_RESULTS_PER_CHALLENGE,
        "results_per_challenge",
    )?;
    if unique_ids.len() * result_limit > MAX_TOTAL_RESULTS {
        bail!("Total requested results cannot exceed {MAX_TOTAL_RESULTS}");
    }
    let page_size = positive_integer(
        object.and_then(|object| object.get("page_size")),
        10,
        MAX_PAGE_SIZE,
        "page_size",
    )?
    .min(result_limit);

    let region = match object.and_then(|object| object.get("region")) {
        None | Some(Value::Null) => None,
        Some(Value::String(value)) => {
            let value = js_trim(value).to_uppercase();
            if value.is_empty() {
                None
            } else if value.len() == 2 && value.bytes().all(|byte| byte.is_ascii_uppercase()) {
                Some(value)
            } else {
                bail!("region must be a two-letter country code");
            }
        }
        Some(_) => bail!("region must be a two-letter country code"),
    };

    let initial_cursor = match object.and_then(|object| object.get("cursor")) {
        None | Some(Value::Null) => None,
        Some(Value::String(value)) => {
            let cursor = js_trim(value);
            (!cursor.is_empty()).then(|| cursor.to_owned())
        }
        Some(value) => match safe_integer_string(value) {
            Some(cursor) => Some(cursor),
            None => bail!("cursor must be a string or safe integer"),
        },
    };

    Ok(unique_ids
        .into_iter()
        .map(|challenge_id| ChallengeRequest {
            challenge_id,
            region: region.clone(),
            initial_cursor: initial_cursor.clone(),
            result_limit,
            page_size,
        })
        .collect())
}

fn normalize_challenge_id(value: &Value) -> Option<String> {
    let normalized = match value {
        Value::String(value) => js_trim(value).to_owned(),
        _ => safe_integer_string(value)?,
    };
    let length = normalized.len();
    (1..=100)
        .contains(&length)
        .then_some(normalized.clone())
        .filter(|value| value.bytes().all(|byte| byte.is_ascii_digit()))
}

fn safe_integer_string(value: &Value) -> Option<String> {
    let number = value.as_f64()?;
    if !number.is_finite() || number.fract() != 0.0 || number.abs() > MAX_SAFE_INTEGER {
        return None;
    }
    Some(format!("{number:.0}"))
}

fn positive_integer(
    value: Option<&Value>,
    fallback: usize,
    maximum: usize,
    name: &str,
) -> Result<usize> {
    let Some(value) = value else {
        return Ok(fallback);
    };
    let valid_number = value.as_f64().filter(|number| {
        number.is_finite() && number.fract() == 0.0 && *number >= 1.0 && *number <= maximum as f64
    });
    match valid_number {
        Some(number) => Ok(number as usize),
        None => bail!("{name} must be an integer between 1 and {maximum}"),
    }
}

fn is_js_whitespace(character: char) -> bool {
    matches!(
        character,
        '\u{0009}'..='\u{000d}'
            | '\u{0020}'
            | '\u{00a0}'
            | '\u{1680}'
            | '\u{2000}'..='\u{200a}'
            | '\u{2028}'
            | '\u{2029}'
            | '\u{202f}'
            | '\u{205f}'
            | '\u{3000}'
            | '\u{feff}'
    )
}

fn js_trim(value: &str) -> &str {
    value.trim_matches(is_js_whitespace)
}
